### PR TITLE
Fixed #4137 - Don't close CURL connection on shutdown

### LINKED ISSUE #4137: Bug: libcurl transport layer doesn't correctly handle `Connection: close` response header
Storage team recently found this when investigating on an RST issue between C++ clients and storage service. We're seeing abnormally high number of RST in TCP connections. After reading the source code, I found libcurl transport layer didn't correctly handle "connection: close" response header.

at Line 872 of `curl.cpp`, the connection is immediately closed the moment we see this response header.

https://github.com/Azure/azure-sdk-for-cpp/blob/121c103d3baf267c12f8943c7cbafcd8c08a3e7d/sdk/core/azure-core/src/http/curl/curl.cpp#L872

But in [RFC2616, section 14.10](https://www.rfc-editor.org/rfc/rfc2616#section-14.10), it says: the connection SHOULD NOT be considered `persistent' **after** the current request/response is complete.

The expected behavior should be: client side reads response body as normal. After that, instead of putting this connection back to the pool, it closes this connection.

### PR DESCRIPTION
**Pull request by LarryOsterman**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (RickWinter) [inline sdk/core/azure-core/src/http/curl/curl.cpp:562]: Sockets allow you to shutdown the socket for sends while keeping the socket open for receives.  Is there a reason we don't want that behavior here?

https://learn.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-shutdown

After your change, Where do we shutdown the socket?

**Comment by Jinming-Hu** [comment]
I fixed this issue in a different way. https://github.com/Azure/azure-sdk-for-cpp/issues/4137#issuecomment-1331858548

I'm not quite familiar with the libcurl transport layer implementation. You decide.

**Comment by Jinming-Hu** [comment]
I don't get it. How does the change in this PR prevent the connection being moved back into the pool?

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (LarryOsterman) [reply sdk/core/azure-core/src/http/curl/curl.cpp:562]: We shut down the socket when the connection is released (via curl_easy_shutdown).

With my change, if a Connection: close is set, we won't put the connection back into the connection pool, it will just be normally freed.

That allows all the normal operations to be performed on the socket, most importantly, if the operation is a "put" operation and the server responds with "Connection: close", if we called shutdown to block sends, then the "put" would fail.

The new behavior is consistent with the RFC mentioned in the original bug report: Connection: close means "don't persist the connection after this HTTP transaction".

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (vhvb1989) [inline sdk/core/azure-core/src/http/curl/curl.cpp:1077]: This used to be there to test how we react after a connection lost.
There was a test like: https://github.com/Azure/azure-sdk-for-cpp/pull/1721/files#diff-34f6f1c9be490d635445c2e8956dc6aa5f3c6c96f96a33ba9b658fa0f766465cR183

Are we replacing that test to actually make a call to `AzureSdkHttpbinServer::Status(101)` which is expected to shutdown a connection?

All of this logic around `shutdown`  is trying to solve this issue: https://github.com/Azure/azure-sdk-for-cpp/issues/1716

I think, at some point, we started to use the `shutdown` method as a way to avoid a connection to be re-used, which is not good.  In that case, removing the `shutdown` method makes total sense, as that was something for testing only, and not expected to be used by the client implementation.

**Review by vhvb1989 — APPROVED** [review]

File: sdk/core/azure-core/test/ut/curl_connection_pool_test.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/curl_connection_pool_test.cpp:604]: Can we have a test that resembles the real-life scenario that this PR is fixing, a bit more?

I don't follow how the PR is fixing the issue, other than returning a `CURL_SEND_ERROR` in one case, in `SendBuffer`.

**Comment by LarryOsterman** [comment]
> I fixed this issue in a different way. [#4137 (comment)](https://github.com/Azure/azure-sdk-for-cpp/issues/4137#issuecomment-1331858548)
> 
> I'm not quite familiar with the libcurl transport layer implementation. You decide.

Calling Shutdown ensures that the connection is not recycled into the connection pool. If you don't ensure that the connection isn't recycled, then it defeats the purpose of `Connection::close`.

File: sdk/core/azure-core/test/ut/curl_connection_pool_test.cpp
Comment (LarryOsterman) [reply sdk/core/azure-core/test/ut/curl_connection_pool_test.cpp:604]: I believe that this test should *exactly* reflect the real-life scenario that this PR is fixing. We call into the azuresdkforcpp test server asking for a 102 response, which also includes a Connection: close response. Before my fix, the call to Perform would have failed, after the fix, the call to Perform succeeds and returns headers indicating that the connection was closed.

Could you help me understand why it doesn't reflect the real-life scenario?

File: sdk/core/azure-core/test/ut/curl_connection_pool_test.cpp
Comment (LarryOsterman) [reply sdk/core/azure-core/test/ut/curl_connection_pool_test.cpp:612]: There is a way we can test the internals of the transport.

We can define "interfaces" (pure virtual base classes) for the underlying socket implementation (ie a CURL interface and a WinHTTP interface) and then mock that interface in our test collateral. We can then call into the transport with the mocked CURL/WinHTTP and have the CURL/WinHTTP behave in whatever fashion we want.

It's a lot of work, but it is completely possible (it will be rather challenging to do this for WinHTTP since we're now using WinHTTP asynchronously, but it is absolutely possible).

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (LarryOsterman) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1068]: Yes - that's why I had to remove it (it was called when a Connection: close header was received).

The problem in this bug is that the `CurlConnection::Shutdown` method shut down the underlying socket, which meant that we didn't allow processing any subsequent "recv" calls, which violated the RFC.

According to the bug, the RFC says that the *only* thing we should be doing in receipt of a `Connection: close` call is to prevent the connection from being re-used, we should NOT hang up the connection.

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (LarryOsterman) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1077]: FWIW, I keep that test working (although I'm suspicious of the test because it may cause "put" and "patch" commands to malfunction if they receive a connection: close header.

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (LarryOsterman) [reply sdk/core/azure-core/src/http/curl/curl.cpp:562]: > I don't get it. How does the change in this PR prevent the connection being moved back into the pool?

When a connection is marked as "Shutdown", it won't be put back into the pool (see `CurlConnectionPool::MoveConnectionBackToPool` which ignores connections marked as `Shutdown`).

What this PR does is to prevent the underlying socket used in the connection from being shut down, which has the effect of allowing the connection to be processed normally (sends and receives continue to work).

Based on the text you quoted from the RFC, all `Connection: close` is supposed to do is to prevent a connection from being re-used, which is what the `CurlConnection::Shutdown` function does.

**Comment by LarryOsterman** [comment]
> I don't get it. How does the change in this PR prevent the connection being moved back into the pool?

When a connection is marked as "Shutdown", it won't be put back into the pool (see `CurlConnectionPool::MoveConnectionBackToPool` which ignores connections marked as `Shutdown`).

What this PR does is to prevent the underlying socket used in the connection from being shut down, which has the effect of allowing the connection to be processed normally (sends and receives continue to work).

Based on the text you quoted from the RFC, all `Connection: close` is supposed to do is to prevent a connection from being re-used, which is what the `CurlConnection::Shutdown` function does.

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1068]: I guess what I am trying to understand is, with this method override gone, what is this call supposed to be doing?
https://github.com/Azure/azure-sdk-for-cpp/blob/b3b2b6774d36bfbe1d0b08174282341152af4a27/sdk/core/azure-core/src/http/curl/curl.cpp#L871-L872

Comment (LarryOsterman) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1068]: Without the virtual override, this calls the shutdown method on the base CurlConnection object.

The shutdown method on the base CurlConnection object just sets the "m_isShutdown" member variable, which tells the connection cache to not cache the connection.

Comment (LarryOsterman) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1068]: Btw, I misspoke. The virtual CurlNetworkConnection::Shutdown called ::shutdown to force the socket closed. THat was the root of the bug. My fix was to remove that force-close.

**Review by Jinming-Hu — APPROVED** [review]

**Comment by Jinming-Hu** [comment]
@LarryOsterman When can we merge this PR? I happen to have an environment that can repro this bug. I want to try it out.

**Comment by ahsonkhan** [comment]
@Jinming-Hu you could try and validate that this PR fixes the issue, without waiting for the PR to be merged. Just checkout Larry's branch:
https://github.com/LarryOsterman/azure-sdk-for-cpp/tree/larryo/Bug4137

**Comment by Jinming-Hu** [comment]
> @Jinming-Hu you could try and validate that this PR fixes the issue, without waiting for the PR to be merged. Just checkout Larry's branch: https://github.com/LarryOsterman/azure-sdk-for-cpp/tree/larryo/Bug4137

Verification of the fix is not an easy job. It may take several hours. Even if I test this PR, I'll still verify after it's merged into `main` branch.

**Comment by antkmsft** [comment]
CI is currently failing on Windows due to #4152.

**Comment by Jinming-Hu** [comment]
I've tested. Now I can confirm #4137 can be fixed with this PR.

**Review by RickWinter — APPROVED** [review]

**Comment by antkmsft** [comment]
@LarryOsterman probably won't get to merge this PR himself soon. After consultations within our team, we decided to merge this PR. Everyone approved it, and @Jinming-Hu has tested it. So I am going to resolve all the comment threads so GitHub allows merging, and will merge this PR.

File: sdk/core/azure-core/test/ut/curl_connection_pool_test.cpp
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/curl_connection_pool_test.cpp:604]: @ahsonkhan, I am going to resolve this comment thread now so that I can merge this PR.
If you'd like more follow up your question, please create a work item so that we can follow up on it later.

File: sdk/core/azure-core/test/ut/curl_connection_pool_test.cpp
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/curl_connection_pool_test.cpp:611]: @ahsonkhan, I am going to resolve this comment thread now so that I can merge this PR.
If you'd like more follow up your question, please create a work item so that we can follow up on it later.

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (antkmsft) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1068]: @ahsonkhan, I am going to resolve this comment thread now so that I can merge this PR.
If you'd like more follow up your question, please create a work item so that we can follow up on it later.

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (antkmsft) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1077]: @vhvb1989, I am going to resolve this comment thread now so that I can merge this PR.
If you'd like more follow up your question, please create a work item so that we can follow up on it later.

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (antkmsft) [reply sdk/core/azure-core/src/http/curl/curl.cpp:562]: @RickWinter, I am going to resolve this comment thread now so that I can merge this PR.
If you'd like more follow up your question, please create a work item so that we can follow up on it later.

**Pull request merged**
